### PR TITLE
Convert uart half_duplex to use user configurable IO

### PR DESCRIPTION
**Pull request by kkoppul2**
Half duplex UART requires the use of Open drain output with a pull up resistor. This modification has been tested and shown to work between two NUCLEOU545 boards. The verbage on the pin configuration is the same across the reference manuals for all of the chip families that support this behavior.

### CODE REVIEW

**Comment by Dirbaio** [comment]
can you make this configurable? Push-pull half-duplex UART still has use cases, such as allowing for faster baudrates. You just have to ensure you never have conflicts, or add a series resistor to limit the current on conflict.

File: embassy-stm32/src/usart/mod.rs [hunk 1]
```diff
@@ -1051,7 +1066,21 @@ impl<'d> Uart<'d, Async> {
         Self::new_inner(
             peri,
             None,
-            new_pin!(tx, AfType::output(OutputType::PushPull, Speed::Medium)),
+            new_pin!(
+                tx,
+                match half_duplex {
```
Comment (Dirbaio) [inline embassy-stm32/src/usart/mod.rs:1071]: this is copypasted 4 times, can you extract it to a method like `HalfDuplexConfig::af_type()`?

File: embassy-stm32/src/usart/mod.rs [hunk 2]
```diff
@@ -211,6 +211,30 @@ impl Default for Config {
...
+    PushPull,
+    /// Open drain output using external pull up resistor
+    OpenDrainExternal,
+    #[cfg(not(gpio_v1))]
+    /// Open drain output using internal pull up resistor
+    OpenDrainInternal,
+}
+
+impl HalfDuplexConfig {
+    pub fn af_type(self) -> gpio::AfType {
```
Comment (Dirbaio) [inline embassy-stm32/src/usart/mod.rs:228]: i'd make this non-pub, there's no reason for the user to call it plus this way you don't need to doc it.

Comment (kkoppul2) [reply embassy-stm32/src/usart/mod.rs:228]: Of course

**Review by Dirbaio — APPROVED** [review]
thanks!

**Pull request merged**